src/resolution/common.rs

```diff
@@ -0,0 +1,248 @@
+  )]
+  DistTagVersionNotFound {
+    package_name: String,
+    dist_tag: String,
+    version: String,
+  },
+  #[error(
+    "Could not find version '{version}' for npm package '{package_name}'."
+  )]
+  VersionNotFound {
+    package_name: String,
+    version: Version,
+  },
+  // todo(dsherret): This shouldn't reference a CLI flag.
+  #[error(
+    "Could not find npm package '{package_name}' matching '{version_req}'. Try retrieving the latest npm package information by running with --reload"
+  )]
+  VersionReqNotMatched {
+    package_name: String,
+    version_req: VersionReq,
+  },
+}
+
+pub fn resolve_best_package_version_info<'info, 'version>(
+  version_req: &VersionReq,
+  package_info: &'info NpmPackageInfo,
+  existing_versions: impl Iterator<Item = &'version Version>,
+) -> Result<&'info NpmPackageVersionInfo, NpmPackageVersionResolutionError> {
+  if let Some(version) = resolve_best_from_existing_versions(
+    version_req,
+    package_info,
+    existing_versions,
+  )? {
+    match package_info.versions.get(version) {
+      Some(version_info) => Ok(version_info),
+      None => Err(NpmPackageVersionResolutionError::VersionNotFound {
+        package_name: package_info.name.clone(),
+        version: version.clone(),
+      }),
+    }
+  } else {
+    // get the information
+    get_resolved_package_version_and_info(version_req, package_info)
+  }
+}
+
+fn get_resolved_package_version_and_info<'a>(
+  version_req: &VersionReq,
+  info: &'a NpmPackageInfo,
+) -> Result<&'a NpmPackageVersionInfo, NpmPackageVersionResolutionError> {
+  if let Some(tag) = version_req.tag() {
+    tag_to_version_info(info, tag)
+  } else {
+    let mut maybe_best_version: Option<&'a NpmPackageVersionInfo> = None;
+    for version_info in info.versions.values() {
+      let version = &version_info.version;
+      if version_req.matches(version) {
+        let is_best_version = maybe_best_version
+          .as_ref()
+          .map(|best_version| best_version.version.cmp(version).is_lt())
+          .unwrap_or(true);
+        if is_best_version {
+          maybe_best_version = Some(version_info);
+        }
+      }
+    }
+
+    match maybe_best_version {
+      Some(v) => Ok(v),
+      // If the package isn't found, it likely means that the user needs to use
+      // `--reload` in the CLI to get the latest npm package information. Although
+      // it seems like we could make this smart by fetching the latest information for
+      // this package here, we really need a full restart. There could be very
+      // interesting bugs that occur if this package's version was resolved by
+      // something previous using the old information, then now being smart here
+      // causes a new fetch of the package information, meaning this time the
+      // previous resolution of this package's version resolved to an older
+      // version, but next time to a different version because it has new information.
```

Comment:
This comment is probably no longer valid in context of this crate

Reply:
I want to keep it for now. I'm going to be changing this in the future.